File: frontend/src/utils/currency.js
/**
 * Currency utilities for frontend
 * Handles formatting and parsing with locale support
 */

export const CURRENCY_CONFIG = {
  EUR: { symbol: '€', decimals: 2, minorUnit: 100 },
  USD: { symbol: '$', decimals: 2, minorUnit: 100 }
};

/**
 * Format amount with locale-aware formatting
 * @param {number} minorUnits - Amount in minor units (cents)
 * @param {string} currencyCode - Currency code (EUR, USD)
 * @param {string} locale - Locale for formatting (en-US, it-IT)
 * @returns {string} Formatted amount
 */
export function formatCurrency(minorUnits, currencyCode = 'EUR', locale = 'en-US') {
  if (minorUnits === null || minorUnits === undefined) return '-';
  
  const config = CURRENCY_CONFIG[currencyCode] || CURRENCY_CONFIG.EUR;
  const decimalAmount = minorUnits / config.minorUnit;
  
  try {
    return new Intl.NumberFormat(locale, {
      style: 'currency',
      currency: currencyCode,
      minimumFractionDigits: config.decimals,
      maximumFractionDigits: config.decimals
    }).format(decimalAmount);
  } catch (error) {
    // Fallback if locale not supported
    return `${config.symbol}${decimalAmount.toFixed(config.decimals)}`;
  }
}

/**
 * Format number with locale-aware formatting
 * @param {number} value - Number to format
 * @param {string} locale - Locale for formatting
 * @param {number} decimals - Decimal places
 * @returns {string} Formatted number
 */
export function formatNumber(value, locale = 'en-US', decimals = 2) {
  if (value === null || value === undefined) return '-';
  
  try {
    return new Intl.NumberFormat(locale, {
      minimumFractionDigits: decimals,
      maximumFractionDigits: decimals
    }).format(value);
  } catch (error) {
    return value.toFixed(decimals);
  }
}

/**
 * Format date with locale-aware formatting
 * @param {string|Date} date - Date to format
 * @param {string} locale - Locale for formatting
 * @returns {string} Formatted date
 */
export function formatDate(date, locale = 'en-US') {
  if (!date) return '-';
  
  try {
    const dateObj = typeof date === 'string' ? new Date(date) : date;
    return new Intl.DateTimeFormat(locale, {
      year: 'numeric',
      month: 'short',
      day: 'numeric'
    }).format(dateObj);
  } catch (error) {
    return date.toString();
  }
}

/**
 * Parse decimal input to minor units
 * @param {string} input - User input (e.g., "8.50" or "8,50")
 * @param {string} currencyCode - Currency code
 * @returns {number} Amount in minor units
 */
export function parseToMinorUnits(input, currencyCode = 'EUR') {
  if (!input) return 0;
  
  const config = CURRENCY_CONFIG[currencyCode] || CURRENCY_CONFIG.EUR;
  
  // Normalize comma/dot for decimal separator
  const normalized = input.toString().replace(',', '.');
  const decimal = parseFloat(normalized);
  
  if (isNaN(decimal)) return 0;
  
  return Math.round(decimal * config.minorUnit);
}

/**
 * Convert minor units to decimal for display in input
 * @param {number} minorUnits - Amount in minor units
 * @param {string} currencyCode - Currency code
 * @returns {string} Decimal string
 */
export function fromMinorUnits(minorUnits, currencyCode = 'EUR') {
  if (minorUnits === null || minorUnits === undefined) return '';
  
  const config = CURRENCY_CONFIG[currencyCode] || CURRENCY_CONFIG.EUR;
  return (minorUnits / config.minorUnit).toFixed(config.decimals);
}

/**
 * Get currency symbol
 * @param {string} currencyCode - Currency code
 * @returns {string} Currency symbol
 */
export function getCurrencySymbol(currencyCode = 'EUR') {
  return CURRENCY_CONFIG[currencyCode]?.symbol || '€';
}

File: backend/app/utils/currency.py
# backend/app/utils/currency.py
from __future__ import annotations
from backend.app.core.config import settings

CURRENCY_CONFIG = {
	"EUR": {"symbol": "€", "decimals": 2, "minorUnit": 100},
	"USD": {"symbol": "$", "decimals": 2, "minorUnit": 100}
}

# Static exchange rates (EUR base)
EXCHANGE_RATES = {
	"EUR": 1.0,
	"USD": 1.08
}


def normalize_minor_units(amount: float, currency: str | None = None) -> int:
	cur = (currency or settings.DEFAULT_CURRENCY).upper()
	decimals = 0 if cur in {"JPY"} else 2
	return int(round(amount * (10 ** decimals), 0))


def to_minor_units(amount: float, currency_code: str = "EUR") -> int:
	config = CURRENCY_CONFIG.get(currency_code, {"minorUnit": 100})
	return int(round(amount * config["minorUnit"]))


def from_minor_units(amount: int, currency_code: str = "EUR") -> float:
	config = CURRENCY_CONFIG.get(currency_code, {"minorUnit": 100})
	return amount / config["minorUnit"]


def format_amount(amount: int, currency_code: str, locale: str = "en-US") -> dict:
	decimal_amount = from_minor_units(amount, currency_code)
	config = CURRENCY_CONFIG.get(currency_code, {"symbol": "$", "decimals": 2})

	# Use locale-specific formatting
	if locale.startswith("it"):
		# Italian format: € 8,50
		formatted = f"{config['symbol']} {decimal_amount:,.{config['decimals']}f}".replace(",", "X").replace(".",
		                                                                                                     ",").replace(
			"X", ".")
	else:
		# English format: $8.50
		formatted = f"{config['symbol']}{decimal_amount:,.{config['decimals']}f}"

	return {
		"value": amount,
		"currency": currency_code,
		"decimal": decimal_amount,
		"formatted": formatted
	}


def parse_decimal_input(input_str: str, currency_code: str = "EUR") -> int:
	# Normalize comma/dot for decimal separator
	normalized = input_str.replace(",", ".")
	try:
		decimal_amount = float(normalized)
		return to_minor_units(decimal_amount, currency_code)
	except ValueError:
		raise ValueError(f"Invalid amount format: {input_str}")


def convert_currency(amount: int, from_currency: str, to_currency: str) -> int:
	if from_currency == to_currency:
		return amount

	# Convert to EUR base
	decimal_amount = from_minor_units(amount, from_currency)
	eur_amount = decimal_amount / EXCHANGE_RATES.get(from_currency, 1.0)

	# Convert to target currency
	target_amount = eur_amount * EXCHANGE_RATES.get(to_currency, 1.0)
	return to_minor_units(target_amount, to_currency)


def get_currency_config(currency_code: str = "EUR") -> dict:
	return CURRENCY_CONFIG.get(currency_code, CURRENCY_CONFIG["EUR"])
